Simplify logging

autayeu opened the pull request:
This pull contains the following changes (in addition to logging simplification itself)
* logging is now standard Maven (no separation of logs to stderr, no shading)
* docs (formatting mostly) improved
* various typos fixed
* updated dependencies (project and plugins, relevant for #216)
* various clean-ups (diamonds, Java 7, try with resources, etc)
* some clean-ups of testing rigs in production code
* thread safety violation via static variable is fixed

ktoso commented on README.md:
```diff
@@ -386,7 +386,7 @@ Start out with with adding the above steps to your project, next paste this **gi
         <property name="buildTime" value="${git.build.time}"/>
         <property name="buildHost" value="${git.build.host}"/>
-        <porperty name="buildVersion" value="${git.build.version}"/>
+        <property name="buildVersion" value="${git.build.version}"/>
```
funny one, thanks (and for all the other typos too, thanks)

ktoso commented on pom.xml:
```diff
@@ -4,40 +4,19 @@
     <groupId>org.sonatype.oss</groupId>
     <artifactId>oss-parent</artifactId>
-    <version>6</version>
+    <version>9</version>
```
:+1:

mvitz replied:
http://central.sonatype.org/pages/apache-maven.html#deprecated-oss-parent

ktoso commented on pom.xml:
```diff
@@ -4,40 +4,19 @@
   <packaging>maven-plugin</packaging>
   <version>2.2.1-SNAPSHOT</version>
-  <name>Git Commit Id Plugin Maven Mojo</name>
+  <name>Git Commit Id Maven Plugin</name>
```
Good idea

ktoso commented on pom.xml:
```diff
@@ -4,40 +4,19 @@
+  <name>Git Commit Id Maven Plugin</name>
   <description>
-    git-commit-id-plugin is a plugin quite similar to
-    https://fisheye.codehaus.org/browse/mojo/tags/buildnumber-maven-plugin-1.0-beta-4 for example but as buildnumber at
```
ok to remove :+1:

ktoso commented on pom.xml:
```diff
@@ -4,40 +4,19 @@
-    }
-
-    Note that the data is exported via maven resource filtering and is really easy to use with spring -
-    which I've explained in detail in this readme https://github.com/ktoso/maven-git-commit-id-plugin
```
Good change, thanks!

ktoso commented on pom.xml:
```diff
@@ -69,16 +48,18 @@
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+
+    <java.target>1.7</java.target>
```
:+1:

ktoso commented on pom.xml:
```diff
@@ -184,6 +161,14 @@
+    </dependency>
   </dependencies>
 
   <build>
```
dependency upgrades LGTM

ktoso commented on pom.xml:
```diff
@@ -198,6 +183,80 @@
+          <exclude>README.md</exclude>
+        </excludes>
+      </testResource>
+    </testResources>
```
neat!

ktoso commented on pom.xml:
```diff
@@ -205,57 +264,34 @@
           <execution>
-            <phase>package</phase>
+            <id>default-descriptor</id>
+            <phase>process-classes</phase>
```
Sounds reasonable. We should document this explicitly in release notes, just in case

ktoso replied:
(task for me during release notes basically, for my reference https://maven.apache.org/guides/introduction/introduction-to-the-lifecycle.html )

ktoso commented on pom.xml:
```diff
@@ -205,57 +264,34 @@
+              <helpPackageName>pl.project13.maven</helpPackageName>
             </configuration>
           </execution>
+-->
```
remove instead of comment out?

autayeu replied:
It might be a fine balance. Commented out code is a ballast (false positives for searches, for example), but can be useful once in a while to avoid searching for a solution. However, it's always a ballast, while a solution - only once in a while and that one might not be the best (since it's already commented out). That tips the scale in favor of removal.

ktoso commented on src/main/java/pl/project13/jgit/DescribeCommand.java:
```diff
@@ -134,7 +117,7 @@ public DescribeCommand withLoggerBridge(LoggerBridge bridge) {
   public DescribeCommand always(boolean always) {
     this.alwaysFlag = always;
-    log("--always =", always);
+    mojo.getLog().info("--always = " + always);
```
Awesome, thanks a lot!

ktoso commented on src/main/java/pl/project13/jgit/DescribeCommand.java:
```diff
@@ -401,15 +384,9 @@ private String createMatchPattern() {
-    return buf.toString();
-  }
-
-  private void log(Object... parts) {
```
could have remained but used the `mojo.getLog()` IMO, but not a big problem; I don't mind the current state

ktoso commented on src/main/java/pl/project13/jgit/DescribeCommand.java:
```diff
@@ -401,15 +384,9 @@ private String createMatchPattern() {
-    loggerBridge.log(parts);
+    return "^refs/tags/\\Q" +
+            matchOption.get().replace("*", "\\E.*\\Q").replace("?", "\\E.\\Q") +
+            "\\E$";
```
ok

ktoso commented on src/main/java/pl/project13/jgit/DescribeResult.java:
```diff
@@ -160,7 +160,7 @@ public String dirtyMarker() {
 
   /**
-   * <p>The (possibly) "g" prefixed <strong>abbriverated</strong> object id of a commit.</p>
+   * <p>The (possibly) "g" prefixed <strong>abbreviated</strong> object id of a commit.</p>
```
Nice catch

ktoso commented on src/main/java/pl/project13/jgit/JGitCommon.java:
```diff
@@ -97,16 +80,16 @@
 
-  public String getClosestTagName(@NotNull LoggerBridge loggerBridge,@NotNull Repository repo){
-    Map<ObjectId, List<DatedRevTag>> map = getClosestTagAsMap(loggerBridge,repo);
+  public String getClosestTagName(@NotNull Repository repo, @NotNull Mojo mojo){
```
We have an effort attempting to make the core of this lib not depend on maven, so gradle/sbt/lein can use it - let's keep the LoggerBridge please, even if it's only backed by the Mojo now,

autayeu replied:
I was thinking about this, but the advantages are not very clear to me. Anyway, then this needs another round of refactoring to reintroduce the bridge.

ktoso commented on src/main/java/pl/project13/jgit/JGitCommon.java:
```diff
@@ -147,29 +130,28 @@ public int compare(Map.Entry<ObjectId, List<DatedRevTag>> m1, Map.Entry<ObjectId
     });
 
-    LinkedHashMap<ObjectId, List<DatedRevTag>> result = new LinkedHashMap<ObjectId, List<DatedRevTag>>();
+    LinkedHashMap<ObjectId, List<DatedRevTag>> result = new LinkedHashMap<>();
```
ok

ktoso commented on src/main/java/pl/project13/jgit/JGitCommon.java:
```diff
@@ -133,7 +116,7 @@ public String getClosestTagCommitCount(@NotNull LoggerBridge loggerBridge,@NotNu
 
   private LinkedHashMap<ObjectId, List<DatedRevTag>> sortByDatedRevTag(Map<ObjectId, List<DatedRevTag>> map) {
-    List<Map.Entry<ObjectId, List<DatedRevTag>>> list = new LinkedList<Map.Entry<ObjectId, List<DatedRevTag>>>(map.entrySet());
+    List<Map.Entry<ObjectId, List<DatedRevTag>>> list = new ArrayList<>(map.entrySet());
```
:+1:

ktoso commented on src/main/java/pl/project13/jgit/JGitCommon.java:
```diff
@@ -267,8 +247,7 @@ protected String trimFullTagName(@NotNull String tagName) {
   public List<RevCommit> findCommitsUntilSomeTag(Repository repo, RevCommit head, @NotNull Map<ObjectId, List<String>> tagObjectIdToName) {
-    RevWalk revWalk = new RevWalk(repo);
-    try {
+    try (RevWalk revWalk = new RevWalk(repo)) {
```
:+1:

ktoso replied:
(So it's AutoClosable, right? (I'm not on a laptop I can check sources now))

autayeu replied:
Yes.

ktoso commented on src/main/java/pl/project13/jgit/JGitCommon.java:
```diff
@@ -339,13 +316,11 @@ protected int distanceBetween(@NotNull Repository repo, @NotNull RevCommit child
 
   private void seeAllParents(@NotNull RevWalk revWalk, RevCommit child, @NotNull Set<ObjectId> seen) throws IOException {
-    Queue<RevCommit> q = newLinkedList();
+    Queue<RevCommit> q = new ArrayDeque<>();
```
:+1:

ktoso commented on src/main/java/pl/project13/maven/git/CommitIdGenerationMode.java:
```diff
@@ -14,13 +14,9 @@
 
-package pl.project13.maven.git.log;
-
-public interface LoggerBridge {
```
Let's keep the bridge for now, and back it with the Mojo; such that we make it easier to rip out the core without depending on maven (the Mojo). Thanks!

ktoso commented on src/main/java/pl/project13/maven/git/GitCommitIdMojo.java:
```diff
@@ -52,24 +54,19 @@
  * @since 1.0
  */
-@SuppressWarnings({"JavaDoc"})
+@Mojo(name = "revision", defaultPhase = LifecyclePhase.INITIALIZE, threadSafe = true)
```
great to move away from these comment based config. I failed once trying to move away in the past, can't remember why. :+1:

ktoso commented on src/main/java/pl/project13/maven/git/GitCommitIdMojo.java:
```diff
@@ -52,24 +54,19 @@
   // these properties will be exposed to maven
   public static final String BRANCH = "branch";
   public static final String DIRTY = "dirty";
+  // only one of the following two will be exposed, depending on the commitIdGenerationMode
```
good comment, thanks :)

ktoso commented on src/main/java/pl/project13/maven/git/GitCommitIdMojo.java:
```diff
@@ -88,331 +85,272 @@
-   * Basically, injecting into all projects may slow down the build and you don't always need this feature.
+   * The Maven Session Object.
+   */
+  @Parameter(property = "session", required = true, readonly = true)
```
property is ok here?

autayeu replied:
ok, removed.

ktoso commented on src/main/java/pl/project13/maven/git/GitCommitIdMojo.java:
```diff
@@ -88,331 +85,272 @@
-   * or the email of the committer etc.
+   * <p>List of properties to exclude from the resulting file.
+   * May be useful when you want to hide {@code 'git.remote.origin.url'} (maybe because it contains your repo password?)
+   * or the email of the committer.</p>
```
thanks for adding the <p>'s here

ktoso commented on src/main/java/pl/project13/maven/git/GitCommitIdMojo.java:
```diff
@@ -88,331 +85,272 @@
-
   @NotNull
-  LoggerBridge loggerBridge = new MavenLoggerBridge(getLog(), true);
+  private final VerboseLog log = new VerboseLogger(this, true);
```
Keep `LoggerBridge` here (interface backed by the mojo)

ktoso commented on src/main/java/pl/project13/maven/git/GitCommitIdMojo.java:
```diff
@@ -88,331 +85,272 @@
+        commitIdGenerationModeEnum = CommitIdGenerationMode.valueOf(commitIdGenerationMode.toUpperCase());
+      } catch (IllegalArgumentException e) {
+        log.warn("Detected wrong setting for 'commitIdGenerationMode'. Falling back to default 'flat' mode!");
+        commitIdGenerationModeEnum = CommitIdGenerationMode.FLAT;
```
OK, let's use the enum :+1:

ktoso commented on src/main/java/pl/project13/maven/git/GitCommitIdMojo.java:
```diff
@@ -540,10 +470,8 @@ private void throwWhenRequiredDirectoryNotFound(File dotGitDirectory, Boolean re
-      return properties = project.getProperties();
     } else {
-      return properties = new Properties(); // that's ok for unit tests
+      return properties = project.getProperties();
```
great if we got rod of the magic flag :+1:

ktoso commented on src/main/java/pl/project13/maven/git/GitDataProvider.java:
```diff
@@ -112,8 +117,11 @@ public void loadGitData(@NotNull Properties properties) throws IOException, Mojo
-      // git.commit.id.abbrev      
+      switch (commitIdGenerationMode) {
+        case FULL: put(properties, GitCommitIdMojo.COMMIT_ID_FULL, getCommitId());
+        default: put(properties, GitCommitIdMojo.COMMIT_ID_FLAT, getCommitId());
```
I'd prefer `case FULL / case FLAT` in case we add a new mode it's easier to spot the need to handle here then here

ktoso commented on src/main/java/pl/project13/maven/git/GitDescribeConfig.java:
```diff
@@ -28,7 +28,7 @@
-   * so think twice before disabeling it.
+   * so think twice before disabling it.
    *
    * @parameter default-value=false
```
hm hm, does this not need to be moved to @Parameter too? It's a nested config object.

autayeu replied:
Is nested config handled differently? I thought it's a parameter anyway...

ktoso commented on src/main/java/pl/project13/maven/git/GitDirLocator.java:
```diff
@@ -80,7 +79,7 @@ private File findProjectGitDirectory() {
     while (basedir != null) {
       File gitdir = new File(basedir, Constants.DOT_GIT);
-      if (gitdir != null && gitdir.exists()) {
+      if (gitdir.exists()) {
```
should be ok, right

ktoso commented on src/main/java/pl/project13/maven/git/GitRepositoryState.java:
```diff
@@ -16,8 +16,6 @@
  */
 package pl.project13.maven.git;
 
-//import org.codehaus.jackson.annotate.JsonWriteNullProperties;
```
:)

ktoso commented on src/main/java/pl/project13/maven/git/GitRepositoryState.java:
```diff
@@ -30,7 +28,6 @@
  *
  * @since 1.0
  */
-//@JsonWriteNullProperties(true)
```
Yeah, I guess we're not really using this as example. Ok to remove

ktoso commented on src/main/java/pl/project13/maven/git/JGitProvider.java:
```diff
@@ -233,8 +216,7 @@ protected void finalCleanUp() {
         .on(repository)
-        .withLoggerBridge(super.loggerBridge)
-        .setVerbose(super.verbose)
+        .withMojo(mojo)
```
setVerbose on describe is gone?

ktoso replied:
i.e. what does verbose do now?

autayeu replied:
Verbose does what name prescribes - switches verbose logging on and off. Here it is removed because DescribeResult has no output - no logger - no verbose.

ktoso commented on src/main/java/pl/project13/maven/git/NativeGitProvider.java:
```diff
@@ -56,14 +55,6 @@ public static NativeGitProvider on(@NotNull File dotGitDirectory, @NotNull Logge
-    super.verbose = verbose;
-    super.loggerBridge.setVerbose(verbose);
-    return this;
-  }
```
ok it's gone I see hm... we sure?

ktoso replied:
Ok, we sure because debug logging.

ktoso commented on src/main/java/pl/project13/maven/git/NativeGitProvider.java:
```diff
@@ -409,34 +399,6 @@ private StringBuilder readStderr(InputStream err) throws IOException {
-//        throw new RuntimeException("Unable to attach to git process!", e);
-//      }
-//      return output;
-//    }
```
thanks

ktoso commented on src/main/java/pl/project13/maven/git/log/VerboseLogger.java:
```diff
@@ -0,0 +1,148 @@
+/**
+ * Bridges logging to standard Maven log adhering to verbosity level.
+ */
+public class VerboseLogger implements VerboseLog {
```
Hm hm, I see this could be our "LoggerBridge" hm...

ktoso replied:
How about using this interface (VerboseLog) everywhere, to not touch Mojo directly in many places?
(And pass around the VerboseLog, instead of the Mojo like it is now on this branch)

autayeu replied:
VerboseLog is in fact a bridge. However, it is "thinner" than it was before. The point of not touching Mojo in many places is a good one. But it needs one more round of refactoring then. I was thinking of making it SLF4J-compatible (thinking of future use as a lib), but then I'd need to copy few bits of code from SLF4J to the bridge and that's not the best solution. Depending on SLF4J is not a very good one either because of shading. So I went with using Maven logging directly. It's a good question how "libbable" the code base is - maybe other build managers introduce much specifics of their own, making reuse of code harder. Which tipped me in favor of using Maven directly.

ktoso commented on src/main/java/pl/project13/maven/git/log/VerboseLogger.java:
```diff
@@ -0,0 +1,148 @@
+
+    @Override
+    public void info(CharSequence content, Throwable error) {
+        if (verbose) {
```
ok, this answers my previous question about verbose semantics now. Looks good.

ktoso commented on src/test/java/pl/project13/jgit/DescribeCommandOptionsTest.java:
```diff
@@ -74,7 +75,7 @@ public void apply_shouldDelegateToAllOptions() throws Exception {
 
     Repository repo = mock(Repository.class);
-    DescribeCommand command = DescribeCommand.on(repo);
+    DescribeCommand command = DescribeCommand.on(repo).withMojo(new GitCommitIdMojo());
```
:+1:

ktoso commented on src/test/java/pl/project13/jgit/DescribeCommandTagsIntegrationTest.java:
```diff
@@ -81,20 +81,21 @@ public void shouldFindLightweightTagWithTagsOptionGiven() throws Exception {
 
-    assertThat(res.toString()).contains("lightweight-tag");
+      assertThat(res.toString()).contains("lightweight-tag");
+    }
```
test changes LGTM, thanks!

ktoso commented on src/test/java/pl/project13/maven/git/FileSystemMavenSandbox.java:
```diff
@@ -57,7 +54,7 @@ public FileSystemMavenSandbox(String rootSandboxPath) {
   @NotNull
   public FileSystemMavenSandbox withParentProject(String parentProjectDirName, String packaging) {
-    parentProject = createProject(new File(rootSandboxPath + FILE_SEPARATOR + parentProjectDirName), packaging);
+    parentProject = createProject(new File(rootSandboxPath + File.separator + parentProjectDirName), packaging);
```
hm hm... technically File.pathSeparator would be better?

ktoso replied:
Meh, I'm tired it seems – `separator` is OK.

ktoso commented on src/test/java/pl/project13/maven/git/FileSystemMavenSandbox.java:
```diff
@@ -140,20 +136,6 @@ private void createChildDirIfRequired() throws IOException {
-    } catch (IOException e) {
-      System.out.println("Unable to delete the directory: " + rootSandboxPath);
-    }
-  }
```
yay! :+1:

ktoso commented on src/test/java/pl/project13/maven/git/FileSystemMavenSandbox.java:
```diff
@@ -162,21 +144,14 @@ public MavenProject getChildProject() {
   private MavenProject createProject(File basedir, String packaging) {
     MavenProject project = new MavenProject();
-    project.setFile(new File(basedir + FILE_SEPARATOR + "pom.xml"));
+    project.setFile(new File(basedir + File.separator + "pom.xml"));
```
Now that I see this I think File.pathSeparator was meant for all of these actually

ktoso replied:
Nope, ignore that comment, separator is OK.

ktoso commented on src/test/java/pl/project13/maven/git/GitCommitIdMojoDirtyFilesTest.java:
```diff
@@ -38,20 +38,21 @@ public void testDetectCleanWorkingDirectory() throws Exception {
     mojo.setAbbrevLength(abbrevLength);
     mojo.setDateFormat(dateFormat);
-    mojo.setVerbose(verbose);
+    mojo.setVerbose(true);
```
please keep as field, easier to find there

ktoso commented on src/test/java/pl/project13/maven/git/GitCommitIdMojoIntegrationTest.java:
```diff
@@ -799,7 +803,7 @@ public void shouldGenerateCommitIdOldFashioned(boolean useNativeGit) throws Exce
+                .create();
     MavenProject targetProject = mavenSandbox.getChildProject();
 
     setProjectToExecuteMojoIn(targetProject);
```
awesome to see cleanup gone :+1:

ktoso commented on src/test/java/pl/project13/maven/git/GitCommitIdMojoTest.java:
```diff
@@ -291,8 +291,8 @@ public void testCraftPropertiesOutputFileWithRelativePath() throws IOException {
-    String targetDir = baseDir.getCanonicalPath() + "/";
-    String generateGitPropertiesFilename = targetDir + "target/classes/git.properties";
+    String targetDir = baseDir.getCanonicalPath() + File.separator;
+    String generateGitPropertiesFilename = targetDir + "target" + File.separator + "classes" + File.separator + "git.properties";
```
ugly, but good idea :+1:

ktoso commented on src/test/java/pl/project13/maven/git/GitDirLocatorTest.java:
```diff
@@ -52,7 +52,9 @@ public void shouldUseTheManuallySpecifiedDirectory() throws Exception {
-      dotGitDir.delete();
+      if (!dotGitDir.delete()) {
+        dotGitDir.deleteOnExit();
+      }
```
Mixed feelings here, probably not harmful here but had production trouble with deleteOnExit some years ago... Maybe we should log a warning instead?

ktoso replied:
Just like we do below in the test

ktoso replied:
(can keep the deleteOnExit)

autayeu replied:
I don't know - both ways seem acceptable: deleteOnExit is in tests (not such a critical piece of code). On the other hand - it's inside "target", so clean`ed anyway. In addition, randomization makes deleteOnExit even less needed for test repetition without cleaning.

ktoso commented on src/test/java/pl/project13/maven/git/GitIntegrationTest.java:
```diff
@@ -18,32 +18,68 @@
+   */
+  private final static Random r = new Random();
+
+  private final static String SANDBOX_DIR = "target" + File.separator + "sandbox" + File.separator;
```
:+1:

ktoso commented on src/test/java/pl/project13/maven/git/GitIntegrationTest.java:
```diff
@@ -18,32 +18,68 @@
+    // generate unique sandbox for this test
+    File sandbox;
+    do {
+      currSandbox = SANDBOX_DIR + "sandbox" + Integer.toString(r.nextInt(Integer.MAX_VALUE));
```
could be even better via using ThreadLocalRandom.current() if we want to enable parallel tests

autayeu replied:
Great comment! Done.

ktoso commented on src/test/java/pl/project13/maven/git/GitIntegrationTest.java:
```diff
@@ -18,32 +18,68 @@
+        FileUtils.deleteDirectory(sandbox);
+      }
+    } catch (IOException e) {
+      System.out.println("Unable to delete sandbox. Scheduling deleteOnExit: " + currSandbox);
```
:+1:

ktoso commented on src/test/java/pl/project13/maven/git/log/MavenLoggerBridgeTest.java:
```diff
@@ -1,47 +0,0 @@
-    verify(logger).info(expectedExplicit);
-  }
-
-}
```
ok

ktoso commented on src/test/java/pl/project13/maven/git/log/StdOutLoggerBridgeTest.java:
```diff
@@ -1,46 +0,0 @@
-    // then, should not have thrown
-  }
-
-}
```
ok

ktoso commented on src/test/java/pl/project13/test/utils/AssertException.java:
```diff
@@ -80,7 +80,7 @@ public void failWithExpectedButGot(@NotNull Class<? extends Throwable> expectedC
     };
 
-    static interface Strategy {
+    interface Strategy {
```
are those still used? unexpected change somewhat

autayeu replied:
I was passing by it and wasn't sure to remove. @Test(expected = ...) does the job, switched to it.

ktoso commented:
Went through the entire PR on wife's notebook during a movie ;-)
I like what I see here a lot, thanks @autayeu!

One thing I'd want to change before merging is the way we now pass around Mojo where we used to pass around LoggerBridge - can we pass around the VerboseLog instead there now? I'd like to keep maven as "close to the surface" as possible. Let me know what you think on that one, let's adjust and merge - thanks a lot!

// And welcome to 2016 :fireworks: ;)

autayeu commented:
Thank you for a thorough review. I've pushed some updates already to address most of your comments. The biggest about the bridge is to address yet.

mvitz commented on pom.xml:
```diff
@@ -69,16 +48,18 @@
 
   <properties>
+    <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+    <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
```
Could be even shortened to <encoding>UTF-8</encoding> if I remember correctly.

mvitz commented on pom.xml:
```diff
@@ -205,57 +264,34 @@
         </configuration>
       </plugin>
-      <!-- we need to create a uber-jar for https://github.com/ktoso/maven-git-commit-id-plugin/issues/196 -->
+
```
If `maven.compiler.source` and `maven.compiler.target` would be set as properties to 1.7 we could remove the complete plugin configuration here.

mvitz commented:
I like most of this PR but (maybe just as a note to myself) this makes the modularisation effort (see #92 for the discussion) even harder as now everything is dependent on maven.

autayeu commented:
Thank you for a thorough review!

I've addressed the comments above. Namely, logging bridges now are improved. In addition, I've refactored exception handling to favor modularization, removed few remaining static collection constructors, added project encoding handling to avoid hard-coded charsets.

It turns out property="session" removed earlier is needed. This case stresses the need for proper integration testing: I only got this one with manual testing. It should have been done by integration tests.

Please, review and merge. Thank you!

autayeu commented:
Is there any other outstanding issues with this branch which we need to consider?

ktoso commented:
I'll need to find time to review again – we don't want to make the modularisation effort harder, as mentioned by @mvitz. Please bear with me :)

autayeu commented:
OK!

autayeu commented:
Konrad, please, can you take a look at this pull request? Thank you!

autayeu commented:
Konrad, what are the chances of seeing this merged?

ktoso commented:
I'll review again today or tomorrow, was traveling and now sick. Thanks for your patience

autayeu commented:
Thank you!

autayeu commented:
Konrad, how do I interpret the situation with this pull? You're saying you'll merge it, but months later it is still not merged.

ktoso commented:
Please don't assume malice where none is intended – i was very sick for a few days and then had work travel, was unable to catch up on all things. Thanks for the ping here (that was all that's needed) – will go over it now and merge if all good :)

ktoso commented:
Minor nitpicks but no reason to delay merging, thanks again for your patience :)

autayeu commented:
Thank you for merging! And for useful review! Now the next step - release ;)

autayeu commented:
By the way, your above comments are not nitpicking - there were few things in the pull which were touched by me "passing by", minor refactorings, which strictly speaking should have been separated from this pull.

ktoso commented on src/main/java/pl/project13/maven/git/log/LoggerBridge.java:
```diff
@@ -17,10 +17,312 @@
+    void setVerbose(boolean verbose);
+
+/*
+    The following borrowed from SLF4J Logger.java
```
I had to check license compatibility... seems we're good though - MIT

ktoso commented on src/main/java/pl/project13/maven/git/log/LoggerBridge.java:
```diff
@@ -17,10 +17,312 @@
+    /**
+     * Log a message at the DEBUG level according to the specified format
+     * and argument.
+     * <p/>
```
FYI - this breaks releases from JDK8. It must not contain empty `<p/>`
